feat: external crash reporter

jpnurmi opened the pull request:
When an external crash reporter is registered, instead of posting minidump + attachments straight to Sentry, Crashpad now dumps them into an envelope, spawns and detaches the external crash reporter process, and passes the envelope as an argument.

Required for:
- https://github.com/getsentry/sentry-native/pull/1303

jpnurmi committed:
WIP

jpnurmi committed:
windows

jpnurmi committed:
mac

jpnurmi committed:
mac: open bundles

jpnurmi committed:
win: fix member order

jpnurmi committed:
linux: minidump

jpnurmi committed:
WIP: FeedbackReport

jpnurmi committed:
Clean up DCHECKs

jpnurmi committed:
Resolve TODO

jpnurmi committed:
windows: fix build

jpnurmi committed:
mac: fix build

jpnurmi committed:
fix pp

jpnurmi commented:
@sentry review

jpnurmi commented on client/crash_report_database.cc:
```diff
@@ -204,6 +206,70 @@ bool CrashReportDatabase::UploadReport::Initialize(const base::FilePath& path,
+  for (const auto& attachment : attachments) {
+    std::string contents;
+    base::FilePath basename = attachment.BaseName();
+#if BUILDFLAG(IS_WIN)
+    constexpr std::wstring_view kInternalPrefix = L"__sentry-";
+#else
+    constexpr std::string_view kInternalPrefix = "__sentry-";
+#endif
+    if (basename.value().rfind(kInternalPrefix, 0) == 0 ||
```
as the name suggests, `_Logging_ReadEntireFile` logs relevant failures

jpnurmi committed:
fix FileOffset format specifier

jpnurmi committed:
/usr/bin/open

jpnurmi committed:
explicit cast because off_t depends on the platform

jpnurmi committed:
doc: FeedbackReport

Co-authored-by: seer-by-sentry[bot] <157164994+seer-by-sentry[bot]@users.noreply.github.com>

jpnurmi committed:
doc: LaunchFeedbackHandler

Co-authored-by: seer-by-sentry[bot] <157164994+seer-by-sentry[bot]@users.noreply.github.com>

jpnurmi commented on handler/handler_main.cc:
```diff
@@ -1139,6 +1155,8 @@ int HandlerMain(int argc,
 #if BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_WIN)
       ,options.wait_for_upload
 #endif
+      ,&options.feedback_handler
+      ,&options.feedback_path
   );
```
not a public general purpose command-line tool

jpnurmi commented on client/crashpad_client.h:
```diff
@@ -125,6 +125,10 @@ class CrashpadClient {
+  //! \param[in] feedback_handler The path to a feedback handler executable that
+  //!     will be executed after a crash has been handled.
+  //! \param[in] feedback_path The path to a feedback envelope that will be
+  //!     filled up and passed as an argument to the feedback handler.
```
the feature will be documented in detail in sentry-docs, not here

jpnurmi commented on client/crash_report_database_win.cc:
```diff
@@ -711,6 +714,38 @@ base::FilePath CrashReportDatabaseWin::DatabasePath() {
 
+bool CrashReportDatabaseWin::LaunchFeedbackHandler(
+    const base::FilePath& feedback_handler,
+    const base::FilePath& feedback_path) {
```
there's no reason to hard-code 32000 here. CreateProcess() will fail and return `ERROR_INVALID_PARAMETER` - and PLOG(ERROR) will print the details

jpnurmi commented on client/crash_report_database_mac.mm:
```diff
@@ -341,6 +344,30 @@ OperationStatus ReportsInDirectory(const base::FilePath& path,
+  }
+
+  return SpawnSubprocess(argv, nullptr, 0, use_path, nullptr);
+}
```
i'll buy /usr/bin/open https://github.com/getsentry/crashpad/pull/131/commits/43bbde93723c853400bfcff6f6b55627e2843cb3

jpnurmi committed:
fix formatting

jpnurmi committed:
argh, you were supposed to write docs

jpnurmi committed:
let feedback handler take ownership

jpnurmi committed:
Rename to Crash Reporter

jpnurmi committed:
mac: fix missed rename

jpnurmi committed:
clean up reports

jpnurmi committed:
allow __sentry-breadcrumbN as attachments

vaind commented on client/crash_report_database.cc:
```diff
@@ -204,6 +206,69 @@ bool CrashReportDatabase::UploadReport::Initialize(const base::FilePath& path,
+        "\n{\"type\": \"attachment\", "
+        "\"length\": %zu, "
+        "\"attachment_type\": \"event.attachment\", "
+        "\"filename\": \"%s\"}\n",
```
We should probably escape filenames. Specifically, any special characters (they're valid in linux/macos filesystems but would produce invalid JSON) and backslash on Windows (valid separator but different meaning in JSON - this one is probably easiest to just replace with forward slash)

vaind commented on client/crash_report_database.cc:
```diff
@@ -204,6 +206,69 @@ bool CrashReportDatabase::UploadReport::Initialize(const base::FilePath& path,
+#endif
+    };
+    if (std::find(kFilter.begin(), kFilter.end(), basename) != kFilter.end() ||
+        !LoggingReadEntireFile(attachment, &contents)) {
```
Maybe worth printing a warning in case the file is not readable/available?

jpnurmi replied:
The "logging" file helpers `LoggingOpen/Read/WriteFile` log errors if the operations fail.

vaind commented on client/crash_report_database_generic.cc:
```diff
@@ -321,6 +326,22 @@ base::FilePath CrashReportDatabaseGeneric::DatabasePath() {
   return base_dir_;
 }
 
+void CrashReportDatabaseGeneric::LaunchCrashReporter(
+    const base::FilePath& crash_reporter,
+    const base::FilePath& crash_envelope) {
+#if BUILDFLAG(IS_LINUX)
+  SpawnSubprocess(
+      {
+          crash_reporter.value(),
+          crash_envelope.value(),
+      },
+      nullptr,
+      -1,
+      !crash_reporter.IsAbsolute(),
+      nullptr);
+#endif
+}
```
As this only works on Linux, should id be moved to the linux-specific file?

jpnurmi replied:
Not sure why it's called "generic", but it's only compiled for Linux and Android:
https://github.com/getsentry/crashpad/blob/3f26be16a95d61eefbee25af8355debc6ec36707/client/CMakeLists.txt#L47-L56

vaind reviewed (commented):
I'm very unfamiliar with the codebase here so just some comments below. Plus, does it make sense to add any tests here?

jpnurmi committed:
drop path lookup for better security

jpnurmi committed:
fix windows build

jpnurmi committed:
Escape attachment filenames in JSON

jpnurmi commented:
> does it make sense to add any tests here?

There are some original unit tests from Crashpad, but they are unmaintained and excluded from the CMake port of the build system. Basically, the only way to test this is with an integration test in sentry-native that spawns the external crash reporter and captures crash and feedback envelopes:

https://github.com/getsentry/sentry-native/blob/1753c54a0910bd5c94dc3f80a88b4be6110b79be/tests/test_integration_crashpad.py#L635

supervacuus reviewed (approved):
This makes all sense.

supervacuus commented:
> There are some original unit tests from Crashpad, but they are unmaintained and excluded from the CMake port of the build system.

We can think about enabling the upstream tests, given that we have added considerable changes to our crashpad fork that no longer maintain compatibility with upstream. Tests allow us to verify whether any of our changes affect (untouched) upstream behavior, and would further provide a basis for isolated unit tests, which is less effort for an envelope writer than going full circle via the consuming integration tests.

However, it isn't necessary to include this in this PR.

jpnurmi committed:
Unpack & merge __sentry-event and __sentry-breadcrumbsN

jpnurmi committed:
Try [[maybe_unused]]

jpnurmi committed:
Switch to mpack

jpnurmi committed:
Fix linking?

jpnurmi committed:
Utilize the JSON lib instead of hand-formatting

jpnurmi committed:
Restore original formatting

jpnurmi committed:
Try fix linking again

jpnurmi committed:
Try CXX

jpnurmi commented:
[msgpack/msgpack-cxx](https://github.com/msgpack/msgpack-c/tree/cpp_master) looked nice at first because it's a header-only library, but then I noticed the size:
```
SLOC    Directory       SLOC-by-Language (Sorted)
77415   msgpack-cxx     cpp=73844,ansic=3571
```

[ludocode/mpack](https://github.com/ludocode/mpack) looks more reasonable:
```
SLOC    Directory       SLOC-by-Language (Sorted)
22304   zlib            ansic=16260,cpp=6020,sh=24
17980   nlohmann-json   cpp=17980
9583    mini_chromium   cpp=8429,python=1154
7840    mpack           ansic=7840
4852    lss             ansic=4852
1839    cpp-httplib     cpp=1839
598     xnu             ansic=598
309     getopt          cpp=309
```

What about the JSON library? Is [nlohmann/json](https://github.com/nlohmann/json) ok or should we look for something slimmer?

jpnurmi committed:
Add missing newlines

jpnurmi committed:
Rename mpack.c to mpack.cc

clang++-19: error: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Werror,-Wdeprecated]

supervacuus commented:
In general, I would love to **cut down dependencies to an absolute minimum**. If there is a way in which we can share between the Native SDK and the `crashpad` repos, I'd vote for it, even if it means to vendor redundantly or create a local fork + submodule.

Even more important than source size is the effect on binary size. C++ code tends to grow binaries dramatically due to the combination of template instantiation, aggressive inlining, and dependencies on the standard library. So, even a slim-looking header-only "modern" C++ library can have severe consequences on binary size. The `crashpad_handler` is big enough as is.

In addition to that - if it is modern C++ - it can also introduce problems for building and deployment since those libraries might depend on features that are not evenly supported by toolchains used downstream (this is a recurring pain and it applies on a finer-grained level than C++ standard version, because particular features of the standard are not bound to appear as implementations in the same toolchain version).

> [msgpack/msgpack-cxx](https://github.com/msgpack/msgpack-c/tree/cpp_master) looked nice at first because it's a header-only library, but then I noticed the size:
> 
> ```
> SLOC    Directory       SLOC-by-Language (Sorted)
> 77415   msgpack-cxx     cpp=73844,ansic=3571
> ```
> 
> [ludocode/mpack](https://github.com/ludocode/mpack) looks more reasonable:
> 
> ```
> SLOC    Directory       SLOC-by-Language (Sorted)
> 22304   zlib            ansic=16260,cpp=6020,sh=24
> 17980   nlohmann-json   cpp=17980
> 9583    mini_chromium   cpp=8429,python=1154
> 7840    mpack           ansic=7840
> 4852    lss             ansic=4852
> 1839    cpp-httplib     cpp=1839
> 598     xnu             ansic=598
> 309     getopt          cpp=309
> ```
> 

Concretely, to the above, `ludocode/mpack` is the MsgPack library we vendor here: https://github.com/getsentry/sentry-native/tree/master/vendor. So, yes, I would be delighted if we use it 😅 . It is fine to vendor it in `crashpad` redundantly. 

> What about the JSON library? Is [nlohmann/json](https://github.com/nlohmann/json) ok, or should we look for something slimmer?

I wonder if we could somehow reuse the existing capabilities of the Native SDK here instead of introducing yet another full-featured JSON implementation, of which we only use a small portion. 

For instance, just thinking out loud: what if you embed MsgPack encoded files from the `crashpad_handler` like you did before (or not even embed just "bundle"), and do the entire decoding of MsgPack and conversion back to JSON when you load it in the feedback UI (by using a Native SDK exposed function that is aware of message-pack encoded attachments)?

This way, you could eliminate the need to introduce all these details to the `crashpad_handler` and use it solely for what it is intended: handling crash snapshots and uploading reports outside the monitored application. **There is significant value in keeping the payloads beyond the minidump opaque for the `crashpad_handler`.**

jpnurmi commented:
Sharing code between sentry-native and crashpad would be useful! I still remember well the duplicate screenshotting code. :sweat: 

While I hesitated to make Crashpad aware of the `__sentry-foo` attachments, this way the external crash reporter can submit the envelope straight to Sentry without any processing. The Sentry Crash Reporter reference GUI might not use the Native SDK, by the way. Thanks to the process separation, we can choose any technology. Currently, the strongest candidate is .NET, because the .NET SDK can AOT-compile self-contained, easy-to-deploy static executables out of the box.

jpnurmi commented:
Hmm, if we had IPC on macOS, we could pass breadcrumbs one by one instead of writing them into files... :thinking:

jpnurmi commented:
Static Crashpad builds with MSVC 2022 (`-DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED_LIBS=OFF -DSENTRY_BACKEND=crashpad`)

### master
```sh
jpnurmi@surface MINGW64 ~/Projects/sentry/sentry-native (master)
$ ls -lah build/master/RelWithDebInfo/
total 49M
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 10:49 ./
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 10:49 ../
-rwxr-xr-x 1 jpnurmi 197610 1.2M Aug 13 10:49 crashpad_handler.exe*
-rw-r--r-- 1 jpnurmi 197610  13M Aug 13 10:49 crashpad_handler.pdb
-rwxr-xr-x 1 jpnurmi 197610  38K Aug 13 10:47 crashpad_wer.dll*
-rw-r--r-- 1 jpnurmi 197610 604K Aug 13 10:47 crashpad_wer.pdb
-rw-r--r-- 1 jpnurmi 197610  42K Aug 13 10:49 minidump.dmp
-rw-r--r-- 1 jpnurmi 197610 2.4M Aug 13 10:49 sentry.lib
-rw-r--r-- 1 jpnurmi 197610 868K Aug 13 10:49 sentry.pdb
-rwxr-xr-x 1 jpnurmi 197610 812K Aug 13 10:49 sentry_example.exe*
-rw-r--r-- 1 jpnurmi 197610 6.8M Aug 13 10:49 sentry_example.pdb
-rwxr-xr-x 1 jpnurmi 197610 760K Aug 13 10:48 sentry_fuzz_json.exe*
-rw-r--r-- 1 jpnurmi 197610 6.9M Aug 13 10:48 sentry_fuzz_json.pdb
-rwxr-xr-x 1 jpnurmi 197610 806K Aug 13 10:49 sentry_screenshot.exe*
-rw-r--r-- 1 jpnurmi 197610 6.9M Aug 13 10:49 sentry_screenshot.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.1M Aug 13 10:48 sentry_test_unit.exe*
-rw-r--r-- 1 jpnurmi 197610 7.5M Aug 13 10:48 sentry_test_unit.pdb
-rw-r--r-- 1 jpnurmi 197610  273 Aug 13 10:49 view-hierarchy.json
```

### feat/feedback-handler
```sh
jpnurmi@surface MINGW64 ~/Projects/sentry/sentry-native (feat/feedback-handler)
$ ls -lah build/feedback-handler/RelWithDebInfo/
total 84M
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 10:52 ./
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 10:52 ../
-rwxr-xr-x 1 jpnurmi 197610 1.6M Aug 13 10:52 crashpad_handler.exe*
-rw-r--r-- 1 jpnurmi 197610  17M Aug 13 10:52 crashpad_handler.pdb
-rwxr-xr-x 1 jpnurmi 197610  38K Aug 13 10:51 crashpad_wer.dll*
-rw-r--r-- 1 jpnurmi 197610 604K Aug 13 10:51 crashpad_wer.pdb
-rw-r--r-- 1 jpnurmi 197610  42K Aug 13 10:52 minidump.dmp
-rw-r--r-- 1 jpnurmi 197610 2.4M Aug 13 10:52 sentry.lib
-rw-r--r-- 1 jpnurmi 197610 868K Aug 13 10:52 sentry.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.1M Aug 13 10:52 sentry_crash_reporter.exe*
-rw-r--r-- 1 jpnurmi 197610  11M Aug 13 10:52 sentry_crash_reporter.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.2M Aug 13 10:52 sentry_example.exe*
-rw-r--r-- 1 jpnurmi 197610  11M Aug 13 10:52 sentry_example.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.1M Aug 13 10:52 sentry_fuzz_json.exe*
-rw-r--r-- 1 jpnurmi 197610  12M Aug 13 10:52 sentry_fuzz_json.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.2M Aug 13 10:52 sentry_screenshot.exe*
-rw-r--r-- 1 jpnurmi 197610  12M Aug 13 10:52 sentry_screenshot.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.4M Aug 13 10:52 sentry_test_unit.exe*
-rw-r--r-- 1 jpnurmi 197610  12M Aug 13 10:52 sentry_test_unit.pdb
-rw-r--r-- 1 jpnurmi 197610  273 Aug 13 10:52 view-hierarchy.json
```

jpnurmi committed:
Drop nlohmann-json in favor of manual JSON formatting

jpnurmi commented:
The sizes look reasonable with `nlohmann-json` replaced by manual JSON formatting:

```sh
jpnurmi@surface MINGW64 ~/Projects/sentry/sentry-native (feat/feedback-handler)
$ ls -lah build/manual-json/RelWithDebInfo/
total 58M
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 12:50 ./
drwxr-xr-x 1 jpnurmi 197610    0 Aug 13 12:50 ../
-rwxr-xr-x 1 jpnurmi 197610 1.3M Aug 13 12:50 crashpad_handler.exe*
-rw-r--r-- 1 jpnurmi 197610  13M Aug 13 12:50 crashpad_handler.pdb
-rwxr-xr-x 1 jpnurmi 197610  38K Aug 13 12:48 crashpad_wer.dll*
-rw-r--r-- 1 jpnurmi 197610 604K Aug 13 12:48 crashpad_wer.pdb
-rw-r--r-- 1 jpnurmi 197610  42K Aug 13 12:50 minidump.dmp
-rw-r--r-- 1 jpnurmi 197610 2.4M Aug 13 12:50 sentry.lib
-rw-r--r-- 1 jpnurmi 197610 868K Aug 13 12:50 sentry.pdb
-rwxr-xr-x 1 jpnurmi 197610 782K Aug 13 12:50 sentry_crash_reporter.exe*
-rw-r--r-- 1 jpnurmi 197610 7.0M Aug 13 12:50 sentry_crash_reporter.pdb
-rwxr-xr-x 1 jpnurmi 197610 842K Aug 13 12:50 sentry_example.exe*
-rw-r--r-- 1 jpnurmi 197610 7.0M Aug 13 12:50 sentry_example.pdb
-rwxr-xr-x 1 jpnurmi 197610 787K Aug 13 12:49 sentry_fuzz_json.exe*
-rw-r--r-- 1 jpnurmi 197610 7.0M Aug 13 12:49 sentry_fuzz_json.pdb
-rwxr-xr-x 1 jpnurmi 197610 836K Aug 13 12:50 sentry_screenshot.exe*
-rw-r--r-- 1 jpnurmi 197610 7.0M Aug 13 12:50 sentry_screenshot.pdb
-rwxr-xr-x 1 jpnurmi 197610 1.1M Aug 13 12:49 sentry_test_unit.exe*
-rw-r--r-- 1 jpnurmi 197610 7.7M Aug 13 12:49 sentry_test_unit.pdb
-rw-r--r-- 1 jpnurmi 197610  273 Aug 13 12:50 view-hierarchy.j
```

jpnurmi committed:
Minimize & flatten third_party/mpack

jpnurmi commented:
Hey @supervacuus, is it ok to merge this? I think it's ready to go, but there were significant changes after the approval. I ended up dropping `nlohmann/json` in favor of manual JSON formatting.

I agree that keeping crashpad unaware of the msg-packed attachments would be nice in principle, but passing them unprocessed to the external crash reporter would force the external crash reporter to special-case crashpad envelopes and do non-trivial pre-processing before being able to submit envelopes to Sentry.
